File: backend/app/schemas/fixed_assets.py
"""Fixed Asset Accounting + Depreciation schemas."""
from __future__ import annotations
from datetime import date, datetime
from decimal import Decimal
from typing import Optional, List
from uuid import UUID
from pydantic import BaseModel, ConfigDict, field_validator

from app.models.fixed_assets import (
    DepreciationMethod, DepreciationFrequency, DepreciationStartRule,
    FAAssetStatus, ScheduleStatus, AssetEventType, DisposalMethod,
    FAIAgentType, FAIRecStatus,
)


# ── Asset Category ─────────────────────────────────────────────────────────────

class FAAssetCategoryCreate(BaseModel):
    category_code: str
    category_name: str
    default_depreciation_method: DepreciationMethod = DepreciationMethod.STRAIGHT_LINE
    default_useful_life_months: Optional[int] = None
    default_salvage_value_pct: Optional[Decimal] = Decimal("0")
    capitalization_threshold: Optional[Decimal] = None
    depreciation_start_rule: DepreciationStartRule = DepreciationStartRule.FIRST_OF_NEXT_MONTH
    asset_account: Optional[str] = None
    accum_depreciation_account: Optional[str] = None
    depreciation_expense_account: Optional[str] = None
    disposal_gain_account: Optional[str] = None
    disposal_loss_account: Optional[str] = None
    impairment_account: Optional[str] = None
    revaluation_reserve_account: Optional[str] = None
    maintenance_integration: bool = False
    is_active: bool = True
    notes: Optional[str] = None


class FAAssetCategoryUpdate(BaseModel):
    category_name: Optional[str] = None
    default_depreciation_method: Optional[DepreciationMethod] = None
    default_useful_life_months: Optional[int] = None
    default_salvage_value_pct: Optional[Decimal] = None
    capitalization_threshold: Optional[Decimal] = None
    depreciation_start_rule: Optional[DepreciationStartRule] = None
    asset_account: Optional[str] = None
    accum_depreciation_account: Optional[str] = None
    depreciation_expense_account: Optional[str] = None
    disposal_gain_account: Optional[str] = None
    disposal_loss_account: Optional[str] = None
    impairment_account: Optional[str] = None
    revaluation_reserve_account: Optional[str] = None
    maintenance_integration: Optional[bool] = None
    is_active: Optional[bool] = None
    notes: Optional[str] = None


class FAAssetCategoryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    category_code: str
    category_name: str
    default_depreciation_method: DepreciationMethod
    default_useful_life_months: Optional[int]
    default_salvage_value_pct: Optional[Decimal]
    capitalization_threshold: Optional[Decimal]
    depreciation_start_rule: DepreciationStartRule
    asset_account: Optional[str]
    accum_depreciation_account: Optional[str]
    depreciation_expense_account: Optional[str]
    disposal_gain_account: Optional[str]
    disposal_loss_account: Optional[str]
    impairment_account: Optional[str]
    revaluation_reserve_account: Optional[str]
    maintenance_integration: bool
    is_active: bool
    notes: Optional[str]
    created_at: datetime


# ── Fixed Asset ────────────────────────────────────────────────────────────────

class FAFixedAssetCreate(BaseModel):
    asset_code: str
    asset_name: str
    asset_category_id: UUID
    asset_type: Optional[str] = None
    serial_no: Optional[str] = None
    asset_tag_no: Optional[str] = None
    physical_asset_id: Optional[UUID] = None
    procurement_reference_type: Optional[str] = None
    procurement_reference_id: Optional[UUID] = None
    supplier_id: Optional[UUID] = None
    acquisition_date: Optional[date] = None
    in_service_date: Optional[date] = None
    capitalization_date: Optional[date] = None
    original_cost: Decimal
    currency: str = "KES"
    exchange_rate: Decimal = Decimal("1")
    local_currency_cost: Decimal
    salvage_value: Decimal = Decimal("0")
    useful_life_months: Optional[int] = None
    depreciation_method: DepreciationMethod = DepreciationMethod.STRAIGHT_LINE
    depreciation_frequency: DepreciationFrequency = DepreciationFrequency.MONTHLY
    declining_balance_rate: Optional[Decimal] = None
    total_production_units: Optional[Decimal] = None
    location: Optional[str] = None
    warehouse_id: Optional[UUID] = None
    plant: Optional[str] = None
    department: Optional[str] = None
    cost_center: Optional[str] = None
    custodian_employee_id: Optional[UUID] = None
    condition_status: Optional[str] = None
    is_legacy_import: bool = False
    legacy_accumulated_depreciation: Optional[Decimal] = None
    notes: Optional[str] = None


class FAFixedAssetUpdate(BaseModel):
    asset_name: Optional[str] = None
    asset_type: Optional[str] = None
    serial_no: Optional[str] = None
    asset_tag_no: Optional[str] = None
    physical_asset_id: Optional[UUID] = None
    in_service_date: Optional[date] = None
    salvage_value: Optional[Decimal] = None
    useful_life_months: Optional[int] = None
    depreciation_method: Optional[DepreciationMethod] = None
    declining_balance_rate: Optional[Decimal] = None
    total_production_units: Optional[Decimal] = None
    location: Optional[str] = None
    warehouse_id: Optional[UUID] = None
    plant: Optional[str] = None
    department: Optional[str] = None
    cost_center: Optional[str] = None
    custodian_employee_id: Optional[UUID] = None
    condition_status: Optional[str] = None
    notes: Optional[str] = None


class FAFixedAssetRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    asset_code: str
    asset_name: str
    asset_category_id: UUID
    category_name: Optional[str] = None
    asset_type: Optional[str]
    serial_no: Optional[str]
    asset_tag_no: Optional[str]
    physical_asset_id: Optional[UUID]
    supplier_id: Optional[UUID]
    acquisition_date: Optional[date]
    in_service_date: Optional[date]
    capitalization_date: Optional[date]
    depreciation_start_date: Optional[date]
    depreciation_end_date: Optional[date]
    original_cost: Decimal
    currency: str
    exchange_rate: Decimal
    local_currency_cost: Decimal
    salvage_value: Decimal
    depreciable_base: Decimal
    useful_life_months: Optional[int]
    depreciation_method: DepreciationMethod
    depreciation_frequency: DepreciationFrequency
    declining_balance_rate: Optional[Decimal]
    total_production_units: Optional[Decimal]
    accumulated_depreciation: Decimal
    net_book_value: Decimal
    last_depreciation_date: Optional[date]
    status: FAAssetStatus
    location: Optional[str]
    plant: Optional[str]
    department: Optional[str]
    cost_center: Optional[str]
    condition_status: Optional[str]
    is_legacy_import: bool
    notes: Optional[str]
    created_at: datetime


# ── Capitalize ─────────────────────────────────────────────────────────────────

class FACapitalizeRequest(BaseModel):
    capitalization_date: date
    in_service_date: Optional[date] = None
    depreciation_start_date: Optional[date] = None
    notes: Optional[str] = None


# ── Transfer ───────────────────────────────────────────────────────────────────

class FATransferRequest(BaseModel):
    effective_date: date
    new_location: Optional[str] = None
    new_plant: Optional[str] = None
    new_department: Optional[str] = None
    new_cost_center: Optional[str] = None
    new_custodian_employee_id: Optional[UUID] = None
    reason: Optional[str] = None


# ── Revaluation ────────────────────────────────────────────────────────────────

class FARevaluationRequest(BaseModel):
    revaluation_date: date
    new_carrying_value: Decimal
    reason: str
    treatment: str = "RESERVE"  # RESERVE or PNL


# ── Impairment ─────────────────────────────────────────────────────────────────

class FAImpairmentRequest(BaseModel):
    impairment_date: date
    impairment_amount: Decimal
    reason: str
    notes: Optional[str] = None


# ── Disposal ───────────────────────────────────────────────────────────────────

class FADisposalRequest(BaseModel):
    disposal_date: date
    disposal_method: DisposalMethod
    sale_proceeds: Decimal = Decimal("0")
    buyer_name: Optional[str] = None
    reason: Optional[str] = None
    notes: Optional[str] = None


class FADisposalRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    asset_id: UUID
    disposal_method: DisposalMethod
    disposal_date: date
    sale_proceeds: Optional[Decimal]
    nbv_at_disposal: Decimal
    cost_at_disposal: Decimal
    accum_depr_at_disposal: Decimal
    gain_loss: Decimal
    reason: Optional[str]
    notes: Optional[str]
    created_at: datetime


# ── Depreciation Schedule ──────────────────────────────────────────────────────

class FADepreciationScheduleRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    asset_id: UUID
    period_start: date
    period_end: date
    scheduled_amount: Decimal
    posted_amount: Optional[Decimal]
    posting_date: Optional[date]
    schedule_status: ScheduleStatus
    opening_nbv: Decimal
    closing_nbv: Decimal
    actual_production_units: Optional[Decimal]
    notes: Optional[str]


# ── Depreciation Posting Run ───────────────────────────────────────────────────

class FADepreciationGenerateRequest(BaseModel):
    asset_ids: Optional[List[UUID]] = None  # None = all active assets
    through_date: date


class FADepreciationPostRequest(BaseModel):
    period_start: date
    period_end: date
    asset_ids: Optional[List[UUID]] = None
    dry_run: bool = False


class FADepreciationPostResult(BaseModel):
    total_assets: int
    total_posted: int
    total_amount: Decimal
    failed: List[str] = []
    dry_run: bool


# ── Asset Event ────────────────────────────────────────────────────────────────

class FAAssetEventRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    asset_id: UUID
    event_type: AssetEventType
    event_date: date
    reference_type: Optional[str]
    amount: Optional[Decimal]
    nbv_before: Optional[Decimal]
    nbv_after: Optional[Decimal]
    reason: Optional[str]
    notes: Optional[str]
    created_at: datetime


# ── Component ──────────────────────────────────────────────────────────────────

class FAComponentCreate(BaseModel):
    component_name: str
    component_cost: Decimal
    salvage_value: Decimal = Decimal("0")
    useful_life_months: Optional[int] = None
    depreciation_method: DepreciationMethod = DepreciationMethod.STRAIGHT_LINE
    in_service_date: Optional[date] = None
    notes: Optional[str] = None


class FAComponentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    parent_asset_id: UUID
    component_name: str
    component_cost: Decimal
    salvage_value: Decimal
    useful_life_months: Optional[int]
    depreciation_method: DepreciationMethod
    in_service_date: Optional[date]
    accumulated_depreciation: Decimal
    net_book_value: Decimal
    is_active: bool
    notes: Optional[str]


# ── Legacy Import ──────────────────────────────────────────────────────────────

class FALegacyImportRow(BaseModel):
    asset_code: str
    asset_name: str
    asset_category_id: UUID
    original_cost: Decimal
    accumulated_depreciation_to_date: Decimal
    in_service_date: date
    depreciation_method: DepreciationMethod = DepreciationMethod.STRAIGHT_LINE
    useful_life_months: Optional[int] = None
    salvage_value: Decimal = Decimal("0")
    cost_center: Optional[str] = None
    location: Optional[str] = None
    notes: Optional[str] = None


class FALegacyImportRequest(BaseModel):
    rows: List[FALegacyImportRow]
    as_of_date: date


class FALegacyImportResult(BaseModel):
    total: int
    succeeded: int
    failed: List[str]


# ── NBV Report ─────────────────────────────────────────────────────────────────

class FANBVReportRow(BaseModel):
    asset_id: str
    asset_code: str
    asset_name: str
    category: str
    cost_center: Optional[str]
    original_cost: Decimal
    accumulated_depreciation: Decimal
    net_book_value: Decimal
    status: str
    as_of_date: date


# ── AI Recommendation ──────────────────────────────────────────────────────────

class FAAIRecRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    asset_id: Optional[UUID]
    agent_type: FAIAgentType
    status: FAIRecStatus
    title: str
    detail: str
    severity: Optional[str]
    action_taken: Optional[str]
    created_at: datetime


class FAAIRecAck(BaseModel):
    status: FAIRecStatus
    action_taken: Optional[str] = None

File: frontend/src/lib/fixed_assets.ts
import axios from "axios";

const API = "/api/v1/fixed-assets";

// ── Enums ──────────────────────────────────────────────────────────────────────

export type DepreciationMethod = "STRAIGHT_LINE" | "DECLINING_BALANCE" | "UNITS_OF_PRODUCTION" | "MANUAL";
export type DepreciationFrequency = "MONTHLY" | "QUARTERLY" | "YEARLY";
export type FAAssetStatus = "DRAFT" | "ACTIVE" | "SUSPENDED" | "IMPAIRED" | "DISPOSED" | "RETIRED" | "ARCHIVED";
export type ScheduleStatus = "PLANNED" | "POSTED" | "REVERSED" | "ADJUSTED";
export type AssetEventType =
  | "ACQUISITION" | "CAPITALIZATION" | "TRANSFER" | "REVALUATION" | "IMPAIRMENT"
  | "REPAIR_CAPITALIZATION" | "DEPRECIATION" | "DISPOSAL" | "WRITEOFF"
  | "RETIREMENT" | "SUSPENSION" | "REACTIVATION" | "LEGACY_IMPORT";
export type DisposalMethod = "SALE" | "RETIREMENT" | "SCRAP" | "WRITEOFF";
export type FAIAgentType = "RISK_MONITOR" | "DEPRECIATION_REVIEW" | "CAPITALIZATION_CHECKER";
export type FAIRecStatus = "PENDING" | "ACKNOWLEDGED" | "ACTIONED" | "DISMISSED";

// ── Interfaces ─────────────────────────────────────────────────────────────────

export interface FAAssetCategory {
  id: string;
  category_code: string;
  category_name: string;
  default_depreciation_method: DepreciationMethod;
  default_useful_life_months: number | null;
  default_salvage_value_pct: number | null;
  capitalization_threshold: number | null;
  depreciation_start_rule: string;
  asset_account: string | null;
  accum_depreciation_account: string | null;
  depreciation_expense_account: string | null;
  is_active: boolean;
  notes: string | null;
  created_at: string;
}

export interface FAFixedAsset {
  id: string;
  asset_code: string;
  asset_name: string;
  asset_category_id: string;
  category_name: string | null;
  asset_type: string | null;
  serial_no: string | null;
  asset_tag_no: string | null;
  acquisition_date: string | null;
  in_service_date: string | null;
  capitalization_date: string | null;
  depreciation_start_date: string | null;
  depreciation_end_date: string | null;
  original_cost: number;
  currency: string;
  local_currency_cost: number;
  salvage_value: number;
  depreciable_base: number;
  useful_life_months: number | null;
  depreciation_method: DepreciationMethod;
  depreciation_frequency: DepreciationFrequency;
  declining_balance_rate: number | null;
  accumulated_depreciation: number;
  net_book_value: number;
  last_depreciation_date: string | null;
  status: FAAssetStatus;
  location: string | null;
  plant: string | null;
  department: string | null;
  cost_center: string | null;
  condition_status: string | null;
  is_legacy_import: boolean;
  notes: string | null;
  created_at: string;
}

export interface FADepreciationScheduleRow {
  id: string;
  asset_id: string;
  period_start: string;
  period_end: string;
  scheduled_amount: number;
  posted_amount: number | null;
  posting_date: string | null;
  schedule_status: ScheduleStatus;
  opening_nbv: number;
  closing_nbv: number;
  actual_production_units: number | null;
}

export interface FAAssetEvent {
  id: string;
  asset_id: string;
  event_type: AssetEventType;
  event_date: string;
  amount: number | null;
  nbv_before: number | null;
  nbv_after: number | null;
  reason: string | null;
  notes: string | null;
  created_at: string;
}

export interface FAComponent {
  id: string;
  parent_asset_id: string;
  component_name: string;
  component_cost: number;
  salvage_value: number;
  useful_life_months: number | null;
  depreciation_method: DepreciationMethod;
  in_service_date: string | null;
  accumulated_depreciation: number;
  net_book_value: number;
  is_active: boolean;
}

export interface FADisposal {
  id: string;
  asset_id: string;
  disposal_method: DisposalMethod;
  disposal_date: string;
  sale_proceeds: number | null;
  nbv_at_disposal: number;
  cost_at_disposal: number;
  accum_depr_at_disposal: number;
  gain_loss: number;
  reason: string | null;
  created_at: string;
}

export interface FAAIRecommendation {
  id: string;
  asset_id: string | null;
  agent_type: FAIAgentType;
  status: FAIRecStatus;
  title: string;
  detail: string;
  severity: string | null;
  action_taken: string | null;
  created_at: string;
}

export interface FADashboardSummary {
  total_assets: number;
  active_assets: number;
  total_nbv: number;
  total_cost: number;
  total_accumulated_depreciation: number;
  pending_depreciation_lines: number;
  ai_pending_recommendations: number;
}

export interface FANBVReportRow {
  asset_id: string;
  asset_code: string;
  asset_name: string;
  category: string;
  cost_center: string | null;
  original_cost: number;
  accumulated_depreciation: number;
  net_book_value: number;
  status: string;
  as_of_date: string;
}

export interface FADeprPostResult {
  total_assets: number;
  total_posted: number;
  total_amount: number;
  failed: string[];
  dry_run: boolean;
}

// ── API Client ─────────────────────────────────────────────────────────────────

const r = <T>(p: Promise<{ data: T }>) => p.then((res) => res.data);

export const faApi = {
  // Dashboard
  getDashboard: () => r<FADashboardSummary>(axios.get(`${API}/dashboard/summary`)),

  // Categories
  getCategories: (activeOnly = true) =>
    r<FAAssetCategory[]>(axios.get(`${API}/categories`, { params: { active_only: activeOnly } })),
  createCategory: (data: Partial<FAAssetCategory>) =>
    r<FAAssetCategory>(axios.post(`${API}/categories`, data)),
  updateCategory: (id: string, data: Partial<FAAssetCategory>) =>
    r<FAAssetCategory>(axios.patch(`${API}/categories/${id}`, data)),

  // Assets
  listAssets: (params?: {
    status?: FAAssetStatus; category_id?: string; cost_center?: string; search?: string;
  }) => r<FAFixedAsset[]>(axios.get(API, { params })),
  getAsset: (id: string) => r<FAFixedAsset>(axios.get(`${API}/${id}`)),
  createAsset: (data: Partial<FAFixedAsset>) => r<FAFixedAsset>(axios.post(API, data)),
  updateAsset: (id: string, data: Partial<FAFixedAsset>) =>
    r<FAFixedAsset>(axios.patch(`${API}/${id}`, data)),

  // Lifecycle actions
  capitalize: (id: string, data: { capitalization_date: string; in_service_date?: string; notes?: string }) =>
    r<FAFixedAsset>(axios.post(`${API}/${id}/capitalize`, data)),
  transfer: (id: string, data: object) =>
    r<FAAssetEvent>(axios.post(`${API}/${id}/transfer`, data)),
  revalue: (id: string, data: object) =>
    r<FAAssetEvent>(axios.post(`${API}/${id}/revalue`, data)),
  impair: (id: string, data: object) =>
    r<FAAssetEvent>(axios.post(`${API}/${id}/impair`, data)),
  dispose: (id: string, data: object) =>
    r<FADisposal>(axios.post(`${API}/${id}/dispose`, data)),

  // Depreciation
  getSchedule: (id: string, statusFilter?: ScheduleStatus) =>
    r<FADepreciationScheduleRow[]>(
      axios.get(`${API}/${id}/depreciation-schedule`, { params: { status_filter: statusFilter } })
    ),
  generateSchedule: (data: { asset_ids?: string[]; through_date: string }) =>
    r<{ assets_processed: number; lines_generated: number }>(
      axios.post(`${API}/depreciation/generate`, data)
    ),
  postDepreciation: (data: {
    period_start: string; period_end: string; asset_ids?: string[]; dry_run?: boolean;
  }) => r<FADeprPostResult>(axios.post(`${API}/depreciation/post`, data)),

  // Events / history
  getEvents: (id: string) => r<FAAssetEvent[]>(axios.get(`${API}/${id}/events`)),

  // Components
  getComponents: (id: string) => r<FAComponent[]>(axios.get(`${API}/${id}/components`)),
  addComponent: (id: string, data: object) =>
    r<FAComponent>(axios.post(`${API}/${id}/components`, data)),

  // Legacy import
  importLegacy: (data: object) =>
    r<{ total: number; succeeded: number; failed: string[] }>(axios.post(`${API}/import-legacy`, data)),

  // Reports
  getNBVReport: (params?: { as_of_date?: string; category_id?: string; cost_center?: string }) =>
    r<FANBVReportRow[]>(axios.get(`${API}/reports/nbv`, { params })),
  getRegister: (status?: FAAssetStatus) =>
    r<FAFixedAsset[]>(axios.get(`${API}/reports/register`, { params: { status } })),

  // AI
  getAIRecs: (status?: FAIRecStatus) =>
    r<FAAIRecommendation[]>(axios.get(`${API}/ai/recommendations`, { params: { status } })),
  runAIAgents: () => r<{ recommendations_generated: number }>(axios.post(`${API}/ai/run-agents`)),
  ackAIRec: (id: string, data: { status: FAIRecStatus; action_taken?: string }) =>
    r<FAAIRecommendation>(axios.patch(`${API}/ai/recommendations/${id}`, data)),
};

// ── Formatting helpers ─────────────────────────────────────────────────────────

export const fmtCurrency = (v: number, currency = "KES") =>
  `${currency} ${v.toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`;

export const STATUS_BADGE: Record<FAAssetStatus, string> = {
  DRAFT:     "bg-gray-100 text-gray-600",
  ACTIVE:    "bg-green-100 text-green-700",
  SUSPENDED: "bg-yellow-100 text-yellow-700",
  IMPAIRED:  "bg-orange-100 text-orange-700",
  DISPOSED:  "bg-red-100 text-red-700",
  RETIRED:   "bg-purple-100 text-purple-700",
  ARCHIVED:  "bg-gray-100 text-gray-500",
};

export const DEPR_METHOD_LABEL: Record<DepreciationMethod, string> = {
  STRAIGHT_LINE:       "Straight-Line",
  DECLINING_BALANCE:   "Declining Balance",
  UNITS_OF_PRODUCTION: "Units of Production",
  MANUAL:              "Manual",
};

export const EVENT_COLOR: Record<AssetEventType, string> = {
  ACQUISITION:          "text-blue-600",
  CAPITALIZATION:       "text-green-600",
  TRANSFER:             "text-indigo-600",
  REVALUATION:          "text-purple-600",
  IMPAIRMENT:           "text-orange-600",
  REPAIR_CAPITALIZATION:"text-teal-600",
  DEPRECIATION:         "text-gray-600",
  DISPOSAL:             "text-red-600",
  WRITEOFF:             "text-red-700",
  RETIREMENT:           "text-red-500",
  SUSPENSION:           "text-yellow-600",
  REACTIVATION:         "text-green-500",
  LEGACY_IMPORT:        "text-cyan-600",
};
